rfd/0159-password-status.md

```diff
@@ -0,0 +1,128 @@
+2. To support any other future feature that might require knowing this information.
+
+To do this, we will extend the [`LocalAuthSecrets`](https://github.com/gravitational/teleport/blob/8c6f6eee44fa39098e4a854e6184082a1bfeeaf9/api/proto/teleport/legacy/types/types.proto#L3302) message by adding a `PasswordState` field:
+
+```proto
+message LocalAuthSecrets {
+  bytes PasswordHash = 1 [(gogoproto.jsontag) = "password_hash,omitempty"];
+  string TOTPKey = 2 [(gogoproto.jsontag) = "totp_key,omitempty"];
+  repeated MFADevice MFA = 5 [(gogoproto.jsontag) = "mfa,omitempty"];
+  WebauthnLocalAuth Webauthn = 6 [(gogoproto.jsontag) = "webauthn,omitempty"];
+
+  // New field
+  PasswordState PasswordState = 7 [(gogoproto.jsontag) = "password_flags,omitempty"];
+}
+
+enum PasswordState {
+  // Unable to tell whether the password has been configured.
+  PASSWORD_STATE_UNKNOWN = 0;
+  // Password is known to be not configured.
+  PASSWORD_STATE_UNSET = 1;
+  // Password is known to be configured.
+  PASSWORD_STATE_SET = 2;
+}
+```
+
+This field will be stored under its own resource key, since there is no top-level local auth secrets object in the database; the password hash, MFA and WebAuthn keys are stored separately. Also, the hash is a plain string, not a structured object, so we can't extend it with another field. The resource key will have the following form: `/web/users/<username>/pwd_state`. The resource will be removed along with other user data upon user deletion, since it is located within the username scope.
+
+`PASSWORD_STATE_UNKNOWN` is deliberately set to 0, since this is the default value when there's no information about the password state in the database.
+
+The state changes to `PASSWORD_STATE_UNSET` whenever another user (or cluster onboarding flow) creates a password reset token for a given user, or removes the password (see [Removing a Password](#removing-a-password)). These are scenarios when the system becomes confident that the user has no password. As a consequence, users created after this change is made will never have their password in an unknown state.
+
+The state changes to `PASSWORD_STATE_SET` whenever user sets/resets their password or successfully signs in using a password. The last scenario, in particular, will allow us to gradually fill in information about existing users that have passwords.
+
+> [!WARNING]
+> The PasswordState flag should _not_ be used for authentication itself. Attempting to shortcut password verification may open a timing attack vulnerability.
+
+### Returning Password State to the Account Settings UI
+
+The old account settings page had separate paths for managing passwords and MFA devices, so there is no "global" endpoint that would return information about all authentication methods. There is only an endpoint that returns a list of MFA devices.
+
+We therefore need to expose a special `/webapi/user/passwordstatus` endpoint that will return the `PasswordStatus` flag value — and only this.
```

Comment:
Would it make sense in the `v1/webapi/sites/<site>/context` endpoint?

Reply:
@zmb3 Hmm. It sounded a bit random to stuff it into the structure that's used on every page, but then, if `authType` is already there, then perhaps `passwordStatus` could also belong.